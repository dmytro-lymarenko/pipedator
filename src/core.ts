export interface ValidationError {
	message: string;
	path: string[]; // ['a', '0', 'test'] which is equivalent to 'a[0].test' of the value
}

export interface ValidationContext {
	value: any;
	path: string[];
}

export interface CreateValidatorOptions {
	validate: (value: any, ctx: ValidationContext) => null | ValidationError;
}

export interface Validator {
	validate: (value: any, ctx?: ValidationContext) => null | ValidationError;
	pipe: (...validators: Validator[]) => Validator;
}

export function pipe(...validators: Validator[]) {
	return createValidator({
		validate: (value, ctx) => {
			// find the first error in pipe
			let i: number = 0;
			let error: ValidationError | null = null;

			while (i < validators.length && error === null) {
				error = validators[i].validate(value, ctx);
				i = i + 1;
			}

			return error;
		},
	});
}

/**
 * Don't use ${value} inside message when value is invalid
 * @param options
 */
export function createValidator(options: CreateValidatorOptions): Validator {
	function validate(value: any, ctx?: ValidationContext) {
		return options.validate(value, ctx || { value, path: [] });
	}

	return {
		validate,
		pipe: (...validators) =>
			pipe(
				createValidator({ validate }),
				...validators
			),
	};
}
