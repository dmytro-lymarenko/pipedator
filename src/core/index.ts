export interface ValidationError {
	message: string;
	path: string;
	children: ValidationError[] | null;
}

export interface ValidationContext {
	value: any;
	path: string[];
	rootValue: any;
}

export interface CreateValidatorOptions {
	validate: (value: any, ctx: ValidationContext) => null | ValidationError;
}

export interface Validator<ValidValue = any> {
	validate: (value: any, ctx?: ValidationContext) => null | ValidationError;
	isValid: (value: any) => value is ValidValue;
}

export function findFirstError(
	getValidator: (index: number) => Validator,
	getValue: (index: number) => any,
	getContext: (index: number) => ValidationContext,
	length: number
): { error: ValidationError | null; index: number } {
	let i: number = 0;
	let error: ValidationError | null = null;

	while (i < length && error === null) {
		error = getValidator(i).validate(getValue(i), getContext(i));
		i = i + 1;
	}

	return { error, index: i - 1 };
}

export function getFirstErrors(
	getValidator: (index: number) => Validator,
	getValue: (index: number) => any,
	getContext: (index: number) => ValidationContext,
	length: number
): ValidationError[] {
	const errors: ValidationError[] = [];
	let i: number = 0;
	let error: ValidationError | null = null;

	do {
		error = getValidator(i).validate(getValue(i), getContext(i));
		if (error) {
			errors.push(error);
		}
		i = i + 1;
	} while (i < length && error !== null);

	return errors;
}

/**
 * Don't use ${value} inside message when value is invalid
 * @param options
 */
export function createValidator<ValidValue = any>(options: CreateValidatorOptions): Validator<ValidValue> {
	function validate(value: any, ctx?: ValidationContext) {
		const context: ValidationContext = ctx || {
			value,
			path: [],
			rootValue: value,
		};

		return options.validate(value, context);
	}

	function isValid(value: any): value is ValidValue {
		return validate(value) === null;
	}

	return {
		validate,
		isValid,
	};
}

export function validate(validator: Validator, value: any) {
	const ctx: ValidationContext = {
		value,
		path: [],
		rootValue: value,
	};

	return validator.validate(value, ctx);
}

export function getCurrentPath(ctx: ValidationContext): string {
	return ctx.path.length === 0 ? '' : ctx.path[ctx.path.length - 1];
}

export interface ValidationRef {
	(ctx: ValidationContext): any;
}

export function ref(path: string[]): ValidationRef {
	return (ctx: ValidationContext) => path.reduce((v, key) => v && v[key], ctx.rootValue);
}
