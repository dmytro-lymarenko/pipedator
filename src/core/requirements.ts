export interface BaseRequirement {
	value: any;
	path: string[];
	message: string;
}

export interface RequirementFactory<R extends Requirement = Requirement> {
	(path: string[], value: any): R;
}

export type Requirement = SingleRequirement | DependenceRequirement | AlternativeRequirement | ConcatenationRequirement;

export interface SingleRequirement extends BaseRequirement {
	type: 'single';
}

export interface DependenceRequirement extends BaseRequirement {
	type: 'dependence';
	requirement: Requirement;
}

export interface AlternativeRequirement extends BaseRequirement {
	type: 'alternative';
	requirements: Requirement[];
}

export interface ConcatenationRequirement extends BaseRequirement {
	type: 'concatenation';
	requirements: Requirement[];
}

export function isRequirement(requirement: Requirement | null): requirement is Requirement {
	return requirement !== null;
}

// Factories

export function singleRequirementFactory(message: string): RequirementFactory<SingleRequirement> {
	return (path, value) => ({
		path,
		value,
		message,
		type: 'single',
	});
}

export function dependenceRequirementFactory(
	message: string,
	requirement: Requirement
): RequirementFactory<DependenceRequirement> {
	return (path, value) => ({
		path,
		value,
		requirement,
		message,
		type: 'dependence',
	});
}

export function alternativeRequirementFactory(
	message: string,
	requirements: Requirement[]
): RequirementFactory<AlternativeRequirement | DependenceRequirement> {
	if (requirements.length < 1) {
		throw new Error('There must be 1 or more requirements');
	}

	if (requirements.length === 1) {
		return dependenceRequirementFactory(message, requirements[0]);
	}

	return (path, value) => ({
		path,
		value,
		message,
		type: 'alternative',
		requirements: requirements.reduce(
			(res, requirement) =>
				// (a || (b || (c && d)) || e) === (a || b || (c && d) || e)
				requirement.type === 'alternative' ? [...res, ...requirement.requirements] : [...res, requirement],
			[] as Requirement[]
		),
	});
}

export function concatenationRequirementFactory(
	message: string,
	requirements: Requirement[]
): RequirementFactory<ConcatenationRequirement | DependenceRequirement> {
	if (requirements.length < 1) {
		throw new Error('There must be 1 or more requirements');
	}

	if (requirements.length === 1) {
		return dependenceRequirementFactory(message, requirements[0]);
	}

	return (path, value) => ({
		path,
		value,
		message,
		type: 'concatenation',
		requirements: requirements.reduce(
			(res, requirement) =>
				// (a && (b && (c || d)) && e) === (a && b && (c || d) && e)
				requirement.type === 'concatenation' ? [...res, ...requirement.requirements] : [...res, requirement],
			[] as Requirement[]
		),
	});
}

// examples:
// const numberRequirementFactory = singleRequirementFactory('Value should be a number');

// function equalToRequirementFactory(validValue: any): RequirementFactory<SingleRequirement> {
// 	return singleRequirementFactory(`Value should equal to ${validValue}`);
// }

// function optionalRequirementFactory(requirement: Requirement): RequirementFactory<AlternativeRequirement> {
// 	const undefinedFactory = equalToRequirementFactory(undefined);

// 	return (path, value) => alternativeRequirementFactory('', [undefinedFactory(path, value), requirement])(path, value);
// }

// const optionalNumberRequirementFactory: RequirementFactory<AlternativeRequirement> = (path, value) =>
// 	optionalRequirementFactory(numberRequirementFactory(path, value))(path, value);

// function someRequirementFactory(requirement: Requirement): RequirementFactory<DependenceRequirement> {
// 	return dependenceRequirementFactory('Some item should follow the requirement', requirement);
// }
