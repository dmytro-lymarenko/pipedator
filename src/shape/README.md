- **Description**

Creates a validator to test a value to have provided shape.

- **How to import**

```typescript
import { shape } from 'pipedator';
// or
import { shape } from 'pipedator/lib/shape';

```
- **Signature**

```typescript
function shape<ValidValue = any>(shape: { [key: string]: Validator }, options?: AbstractShapeOptions, message?: string): Validator<ValidValue>;
```
- **Parameters**

  - `shape` - an object in which each key is a validator to test the respective sub value from `value` by that key.
  - `options` - (optional) options (see [`abstractShape`](#abstractshape) for more details)
  - `message` - (optional) custom message.

- **Usage**

```typescript
import { number, string, shape } from 'pipedator';

shape({ a: number(), b: string() }).validate({ a: 1, b: 'text' }); // valid
shape({ a: number(), b: string() }).validate({ a: 1, b: 'text', c: {} }); // valid
shape({ a: number(), b: string() }).validate({ a: '1', b: 'text' }); // invalid, value.a is not a number
shape({ a: number(), b: string() }).validate({ a: 1 }); // invalid, value.b is not a string
shape({ a: number(), b: string() }, { onlyFirstError: true }).validate({ a: '1' }); // invalid, returns only one error
shape({ a: number(), b: string() }, undefined, 'Custom message').validate({ a: 1 }); // invalid with 'Custom message'
```