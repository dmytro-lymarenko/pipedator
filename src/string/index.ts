import { test } from '../test';

export function string<ValidValue = any>(message?: string) {
	return test<ValidValue>(value => typeof value === 'string', message || 'Value should be a string');
}
