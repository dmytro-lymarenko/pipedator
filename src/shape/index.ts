/**
 * @module validators
 */

import { Validator } from '../core';
import { abstractShape } from '../abstractShape';

export function shape(shape: { [key: string]: Validator }, options?: { onlyFirstError?: boolean }, message?: string) {
	const keys = Object.keys(shape);

	return abstractShape(keys, key => shape[key], options, message);
}
