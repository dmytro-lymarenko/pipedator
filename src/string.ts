import { createValidator } from './core';

export function string() {
	return createValidator({
		message: 'Value should be a string',
		validate: (value, ctx) => (typeof value === 'string' ? null : { value, message: ctx.message, path: ctx.path }),
	});
}

export function empty() {
	return createValidator({
		message: 'Value should be empty',
		validate: (value, ctx) => (value.length === 0 ? null : { value, message: ctx.message, path: ctx.path }),
	});
}
