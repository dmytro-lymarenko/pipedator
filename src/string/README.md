- **Description**

Creates a validator to check if a value is a string or not.

- **How to import**

```typescript
import { string } from 'pipedator';
// or
import { string } from 'pipedator/lib/string';

```
- **Signature**

```typescript
function string<ValidValue = any>(message?: string): Validator<ValidValue>;
```
- **Parameters**

  - `message` - (optional) custom message.


- **Usage**

```typescript
string().validate('Some text'); // valid
string().validate(5); // invalid
string('Custom message').validate(null); // invalid with 'Custom message'
```