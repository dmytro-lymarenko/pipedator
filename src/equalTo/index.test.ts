import { equalTo } from './index';
import { ref } from '../core';
import { string } from '../string';
import { shape } from '../shape';

describe('equalTo()', () => {
	it('should validate when value is equal to validValue', () => {
		expect(equalTo('test').validate('test')).toBeNull();
	});

	it('should validate when value is equal to validValue which is a reference', () => {
		expect(shape({ a: string(), b: equalTo(ref(['a'])) }).validate({ a: 'test', b: 'test' })).toBeNull();
	});

	it('should fail when value is not equal to validValue which is a reference', () => {
		expect(shape({ a: string(), b: equalTo(ref(['a'])) }).validate({ a: 'test', b: 'text' })).toMatchSnapshot();
	});

	it('should fail when value is not equal to validValue', () => {
		expect(equalTo('test').validate('text')).toMatchSnapshot();
	});

	it('should support custom message', () => {
		expect(equalTo(1, 'Custom message').validate('test')).toMatchSnapshot();
	});
});
