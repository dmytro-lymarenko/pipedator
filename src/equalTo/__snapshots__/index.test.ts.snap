// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`equalTo() should fail when value is not equal to validValue 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Value should equal to test",
      "path": Array [],
      "value": "text",
    },
  ],
  "rootValue": "text",
}
`;

exports[`equalTo() should fail when value is not equal to validValue which is a reference 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Value should equal to test",
      "path": Array [
        "b",
      ],
      "value": "text",
    },
  ],
  "rootValue": Object {
    "a": "test",
    "b": "text",
  },
}
`;

exports[`equalTo() should support custom message 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Custom message",
      "path": Array [],
      "value": "test",
    },
  ],
  "rootValue": "test",
}
`;
